service/history/historyEventNotifier.go

```diff
@@ -39,15 +41,19 @@ const (
 
 type (
 	historyEventNotifierImpl struct {
+		metrics metrics.Client
+		// internal status indicator
 		status int32
 		// stop signal channel
 		closeChan chan bool
 		// this channel will never close
 		eventsChan chan *historyEventNotification
+		// function which calculate the shard ID from given workflow ID
+		workflowIDToShardID func(string) int
 
-		sync.Mutex
-		// map of workflow identifier to map of subscriber ID to channel
-		eventsPubsubs map[workflowIdentifier]map[string]chan *historyEventNotification
+		metadataLock       sync.RWMutex
+		shardLocks         map[int]*sync.RWMutex
```

Comment:
Mostly the logic looks good, but I'm curious if this can be abstracted out in a separate type?  We implemented something similar with cherami here: https://github.com/uber/cherami-server/blob/master/common/concurrentmap.go.

Reply:
Take a look at this, wrote this cherami. Might be a better fit here https://github.com/uber/cherami-server/blob/master/services/controllerhost/lockmgr.go

Reply:
separate concurrent hash map into common package